nixos/modules/virtualisation/nixos-containers.nix

```
@@ -939,7 +940,16 @@ in
             optional (cfg.networkNamespace != null && (cfg.privateNetwork || cfg.interfaces != [ ]))
               "containers.${name}.networkNamespace is mutally exclusive to containers.${name}.privateNetwork and containers.${name}.interfaces.";
         in
-        mkMerge (mapAttrsToList mapper config.containers);
+        [
+          (mkMerge (mapAttrsToList mapper config.containers))
+          {
+            assertion =
+              (lib.filterAttrs (
+                n: v: (lib.strings.hasPrefix "container@" n) && v.wantedBy != [ ] && !config.boot.enableContainers
+              ) config.systemd.services) == { };
```

I think that since this assertion was based on the misunderstanding of the workflow being talked about, we can drop it as being a very marginal case.

But if we were to keep it, maybe:

```suggestion
            assertion =
              !config.boot.enableContainers ->
              !(any (lib.strings.hasPrefix "container@") (attrNames config.systemd.services));
```

This is simpler and should be more efficient, although I don’t think the perf here will be a big deal since *n* is small. I don’t think we need to condition on `wantedBy`, but I think we’d also need to catch `systemd.targets.*.wants = [ "container@foo.service" ];`. Again, though, I think we can skip this.